README
NAME
    Paludis::UseCleaner - Remove cruft from your use.conf

VERSION
    version 0.01000308

SYNOPSIS
    This module handles the core behaviour of the Use Cleaner, to be
    consumed inside other applications.

    For a "Just Use it" interface, you want paludis-usecleaner.pl and
    Paludis::UseCleaner::App

        my $cleaner = Paludis::UseCleaner->new(
            input     => somefd,
            output    => somefd,
            rejects   => somefd,
            debug     => fd_for_debugging
            dot_trace => fd_for_dot_traces,
          ( # Optional
            display_ui => $object_to_handle_debug_messages
            display_ui_class => $classname_to_construct_a_display_ui
            display_ui_generator => $coderef_to_generate_object_for_display_ui
          )
        );

        $cleaner->do_work();

METHODS
  do_work
        $cleaner->do_work();

    Executes the various transformations and produces the cleaned output
    from the input.

ATTRIBUTES
  input
        $cleaner->input( \*STDIN );
        $cleaner->input( $read_fh );

  output
        $cleaner->output( \*STDOUT );
        $cleaner->output( $write_fh );

  rejects
        $cleaner->rejects( \*STDOUT );
        $cleaner->rejects( $write_fh );

  debug
        $cleaner->debug( \*STDERR );
        $cleaner->debug( $write_fh );

  dot_trace
        $cleaner->dot_trace( \*STDERR );
        $cleaner->dot_trace( $write_fh );

  display_ui
        $cleaner->display_ui( $object );

  display_ui_class
        $cleaner->display_ui_class( 'Some::Class::Name' );

  display_ui_generator
        $cleaner->display_ui_generator( sub {
            my $self = shift;
            ....
            return $object;
        });

PRIVATE METHODS
  __tokenize
        my @line = __tokenize( $line );

    STRIPPED: This method is made invisible to outside code after compile.

  __is_empty_line
        if( __is_empty_line(@line) ){ }

    STRIPPED: This method is made invisible to outside code after compile.

  __is_star_rule
        if( __is_star_rule(@line) ){ }

    STRIPPED: This method is made invisible to outside code after compile.

  __tokenparse
        my ( $spec, $use, $extras ) = __tokenparse( @line );

    STRIPPED: This method is made invisible to outside code after compile.

  __extract_flags
        my ( @flags ) = __extract_flags( \@tokens );

    STRIPPED: This method is made invisible to outside code after compile.

  __extract_label
        my ( $label ) = __extract_label( \@tokens );

    STRIPPED: This method is made invisible to outside code after compile.

  _build_display_ui_class
        my $class = $cleaner->_build_display_ui_class();

  _build_display_ui_generator
        my $generator  $cleaner->_build_display_ui_generator();

  _build_display_ui
        my $object = $cleaner->_build_display_ui();

AUTHOR
    Kent Fredric <kentnl@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Kent Fredric <kentnl@cpan.org>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Paludis-UseCleaner

0.01000308 2013-02-05T08:07:01Z

0.01000307 2010-12-14 17:24:03 Pacific/Auckland
          First version.



Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "File::Find" => 0,
    "File::Temp" => 0,
    "Module::Build" => "0.4003",
    "Test::More" => "0.98"
  },
  "configure_requires" => {
    "Module::Build" => "0.4003"
  },
  "dist_abstract" => "Remove cruft from your use.conf",
  "dist_author" => [
    "Kent Fredric <kentnl\@cpan.org>"
  ],
  "dist_name" => "Paludis-UseCleaner",
  "dist_version" => "0.01000308",
  "license" => "perl",
  "module_name" => "Paludis::UseCleaner",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Cave::Wrapper" => 0,
    "Class::Load" => "0.06",
    "Getopt::Lucid" => 0,
    "IO::Handle" => 0,
    "Moose" => 0,
    "Moose::Util::TypeConstraints" => 0,
    "MooseX::Has::Sugar" => 0,
    "MooseX::Types::Moose" => 0,
    "MooseX::Types::Perl" => 0,
    "Term::ANSIColor" => 0,
    "namespace::autoclean" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [
    "bin/paludis-usecleaner.pl"
  ]
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Paludis/UseCleaner/App.pm

use strict;
use warnings;

package Paludis::UseCleaner::App;
BEGIN {
  $Paludis::UseCleaner::App::AUTHORITY = 'cpan:KENTNL';
}
{
  $Paludis::UseCleaner::App::VERSION = '0.01000308';
}

# ABSTRACT: Command Line App Interface to Paludis::UseCleaner


use Getopt::Lucid qw( :all );

sub _gen_spec {
  return (
    Param(
      "conf|c",
      sub {
        return 1 if $_ eq q{-};
        return 1 if -e $_ && -f $_;
        return;
      }
      )->default("/etc/paludis/use.conf"),
    Param("output|o")->default("/tmp/use.conf.out"),
    Param("rejects|r")->default("/tmp/use.conf.rej"),
    Switch("clobber-output|x")->default(1),
    Switch("clobber-rejects|y")->default(1),
    Switch("quiet|q")->default(1),
    Switch("silent|s")->default(0),
    Switch("help|h")->anycase(),
  );
}

sub _help {
  my @spec = @_;
  my %doc  = (
    'conf'           => {},
    'output'         => {},
    'rejects'        => {},
    'clobber-output' => {},
    'quiet'          => {},
    'silent'         => {},
    'help'           => {},
  );

  for my $rule (@spec) {
    my $name     = $rule->{canon};
    my $doc      = $doc{$name};
    my @switches = split /\|/, $rule->{name};
    for (@switches) {
      if ( ( length $_ ) < 2 ) {
        $_ =~ s/^/-/;
      }
      else {
        $_ =~ s/^/--/;
      }
    }
    @switches = sort { ( length $a ) <=> ( length $b ) } @switches;
    ## no critic(ProhibitComplexMappings)
    if ( $rule->{type} eq 'parameter' ) {
      @switches = map { ( "$_ \$x", "$_=\$x" ) } @switches;
    }
    elsif ( $rule->{type} eq 'switch' ) {
      @switches = map {
        my $i = $_;
        my $j = $i;
        $j =~ s/^--/--no-/;
        ( $j eq $i ) ? $i : ( $i, $j );
      } @switches;
    }
    printf qq{%-50s => %s \n}, ( join q{ }, @switches ), ( defined $rule->{default} ? $rule->{default} : 'undef' );

  }
  return;
}

sub _read_fd {
  my $source = shift;
  ## no critic ( ProhibitPunctuationVars )
  open my $fh, '<', $source or die "Cant open $source $@ $? $!\n";
  return $fh;
}

sub _write_fd {
  my $source = shift;
  ## no critic ( ProhibitPunctuationVars )
  open my $fh, '>', $source or die "Cant open $source $@ $? $!\n";
  return $fh;
}

sub _silent_args {
  return (
    show_skip_empty => 0,
    show_skip_star  => 0,
    show_dot_trace  => 0,
    show_clean      => 0,
    show_rules      => 0,
  );
}

sub _quiet_args {
  return (
    show_skip_empty => 0,
    show_skip_star  => 0,
    show_dot_trace  => 1,
    show_clean      => 0,
    show_rules      => 0,
  );

}

sub _noisy_args {
  return (
    show_skip_empty => 1,
    show_skip_star  => 1,
    show_dot_trace  => 0,
    show_clean      => 1,
    show_rules      => 1,
  );

}

sub run {
  my @spec = _gen_spec();

  my $got = Getopt::Lucid->getopt( \@spec );
  if ( $got->get_help ) {
    _help(@spec);
    exit;

  }

  my %flags = ();

  if ( $got->get_conf eq q{-} ) {
    $flags{input} = \*STDIN;
  }
  else {
    $flags{input} = _read_fd( $got->get_conf );
  }

  if ( $got->get_output eq q{-} ) {
    $flags{output} = \*STDOUT;
  }
  else {
    if ( -e $got->get_output && !$got->get_clobber_output ) {
      die $got->output . " Exists and --no-clobber-output is specified\n";
    }
    $flags{output} = _write_fd( $got->get_output );
  }

  if ( $got->get_rejects eq q{-} ) {
    $flags{rejects} = \*STDERR;
  }
  else {
    if ( -e $got->get_rejects && !$got->get_clobber_rejects ) {
      die $got->get_rejects . " Exists and --no-clobber-rejects is specified\n";
    }
    $flags{rejects} = _write_fd( $got->get_rejects );
  }

  $flags{debug}     = \*STDERR;
  $flags{dot_trace} = \*STDERR;

  ## no critic (ProhibitMagicNumbers)

  $flags{display_ui_generator} = sub {
    my $self = shift;
    require Class::Load;
    Class::Load->VERSION(0.06);
    Class::Load::load_class( $self->display_ui_class );
    return $self->display_ui_class->new(
      do {
        if    ( $got->get_silent ) { _silent_args() }
        elsif ( $got->get_quiet )  { _quiet_args() }
        else                       { _noisy_arg() }
      },
      fd_debug     => $self->debug,
      fd_dot_trace => $self->dot_trace,
    );

  };
  require Class::Load;
  Class::Load->VERSION(0.06);
  Class::Load::load_class('Paludis::UseCleaner');
  my $cleaner = Paludis::UseCleaner->new( \%flags );

  return $cleaner->do_work();
}

1;

__END__

=pod

=head1 NAME

Paludis::UseCleaner::App - Command Line App Interface to Paludis::UseCleaner

=head1 VERSION

version 0.01000308

=head1 SYNOPSIS

This is really just a huge wrapper around L<Getopt::Lucid>
which sets up L<Paludis::UseCleaner> in a friendly way.

    @ARGV=qw( --command -l -i -n -e --arguments );
    use Paludis::UseCleaner::App;

    Paludis::UseCleaner::App->run();

=head1 METHODS

=head2 run

Execute the code.

=head1 COMMAND LINE ARGUMENTS

=head2 --output $file

Set the file to write the cleaned use.conf to.

Defaults as C</tmp/use.conf.out>.

Use C<-> for C<STDOUT>.

=head2 --rejects $file

Set the file to write the rejected lines to.

Defaults as C</tmp/use.conf.rej>.

Use C<-> for C<STDERR>

=head2 --conf

Sets the file to read use.conf from,

Defaults as C</etc/paludis/use.conf>

Use C<-> for C<STDIN>

=head2 --no-clobber-output

If C<--output> exists, die instead of overwriting it.

=head2 --no-clobber-rejects

If C<--rejects> exists, die instead of overwriting it.

=head2 --silent

Print nothing debug related to stderr.

=head2 --no-quiet

Print verbose messages to stderr instead of a dot-trace

=head2 --help

Show a brief command line summary.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Paludis/UseCleaner/ConsoleUI.pm
use strict;
use warnings;

package Paludis::UseCleaner::ConsoleUI;
BEGIN {
  $Paludis::UseCleaner::ConsoleUI::AUTHORITY = 'cpan:KENTNL';
}
{
  $Paludis::UseCleaner::ConsoleUI::VERSION = '0.01000308';
}

# ABSTRACT: SubSpace for handling progress formatting of the cleaner.

use Moose;
use MooseX::Types::Moose qw( :all );
use MooseX::Types::Perl qw( :all );
use MooseX::Has::Sugar;
use IO::Handle;
use Term::ANSIColor;


has show_skip_empty => ( isa => Bool, rw, default => 1 );


has show_skip_star => ( isa => Bool, rw, default => 1 );


has show_dot_trace => ( isa => Bool, rw, default => 0 );


has show_clean => ( isa => Bool, rw, default => 1 );


has show_rules => ( isa => Bool, rw, default => 1 );


has fd_debug => ( isa => GlobRef, rw, required );


has fd_dot_trace => ( isa => GlobRef, rw, required );

my $format = "%s%s\n >> %s\n >  %s%s\n";


sub _message {
  my ( $self, $colour, $label, $line, $reason ) = @_;
  $line =~ s/\n?$//;
  return $self->fd_debug->printf( $format, color($colour), $label, $line, $reason, color('reset') );

}


sub skip_empty {
  my ( $self, $lineno, $line ) = @_;
  $self->dot_trace('>');
  return unless $self->show_skip_empty;
  return $self->_message( 'red', "Skipping $lineno", $line, "Looks empty" );
}


sub skip_star {
  my ( $self, $lineno, $line ) = @_;
  $self->dot_trace('>');
  return unless $self->show_skip_star;
  return $self->_message( 'red', "Skipping $lineno", $line, "* rule" );
}


sub dot_trace {
  my ( $self, $symbol ) = @_;
  return unless $self->show_dot_trace;
  $symbol ||= q{.};
  return $self->fd_dot_trace->print($symbol);
}


sub nomatch {
  my ( $self, $lineno, $line ) = @_;
  $self->dot_trace(q{?});
  return unless $self->show_clean;
  return $self->_message( 'green', "Cleaning $lineno", $line, 'No matching specification' );
}


sub full_rule {
  my ( $self, $spec, $use, $extras ) = @_;
  return unless $self->show_rules;
  $extras->{'use'} = $use;
  my @extradata = map { ( sprintf "%s = [ %s ]", $_, ( join q{, }, @{ $extras->{$_} } ) ) } keys %{$extras};
  return $self->fd_debug->printf( "RULE: spec = $spec %s\n", ( join q{ }, @extradata ) );

}

no Moose;
__PACKAGE__->meta->make_immutable;

1;

__END__

=pod

=head1 NAME

Paludis::UseCleaner::ConsoleUI - SubSpace for handling progress formatting of the cleaner.

=head1 VERSION

version 0.01000308

=head1 METHODS

=head2 skip_empty

    $ui->skip_empty( $lineno, $line )

Notifies user a line has been skipped in the input due to it being empty.
This line has been copied to the output ( cleaned ) file.

=head2 skip_star

    $ui->skip_star( $lineno, $line )

Notifies user a line has been skipped in the input due to it being a * rule,
and thus having far too many possible matches to compute respectably.

This line has been copied to the output ( cleaned ) file.

=head2 dot_trace

    $ui->dot_trace( $symbol = '.' )

Prints a simple progress indicator when show_dot_trace is enabled.

=head2 nomatch

    $ui->nomatch( $lineno, $line )

Notifies use that a line appeared to contain a rule and that rule matched
no packages that exist, both uninstalled and installed, and is thus being removed from the output( cleaned ) file.

Just In case, this line is also copied to the rejects file.

=head2 full_rule

    $extrasmap{VIDEO_CARDS} = \@cardlist;

    $ui->full_rule( $spec, \@useflags, \%extrasmap )

Produces a debug tracing line showing the parsed result of the line as we perceive it internally.

=head1 ATTRIBUTES

=head2 show_skip_empty

    $ui->show_skip_empty( 1 ); # enable showing the empty-line debug
    $ui->show_skip_empty( 0 ); # disable ...

B<default> is C<true>

=head2 show_skip_star

    $ui->show_skip_star( 1 ); # enable showing the * rule debug
    $ui->show_skip_star( 0 ); # disable ...

B<default> is C<true>

=head2 show_dot_trace

    $ui->show_dot_trace( 1 ); # enable showing the dot_trace's
    $ui->show_dot_trace( 0 ); # disable ...

B<default> is C<false>

=head2 show_clean

    $ui->show_clean( 1 ); # enable showing the clean notice
    $ui->show_clean( 0 ); # disable ...

B<default> is C<true>

=head2 show_rules

    $ui->show_rules( 1 ); # enable showing the rule debug
    $ui->show_rules( 0 ); # disable ...

B<default> is C<true>

=head2 fd_debug

    $ui->fd_debug( \*STDOUT ); # debug to stdout
    $ui->fd_debug( $fh ); # debug to a filehandle

=head2 fd_dot_trace

    $ui->fd_dot_trace( \*STDOUT ); # debug to stdout
    $ui->fd_dot_trace( $fh ); # debug to a filehandle

=head1 PRIVATE METHODS

=head2 _message

   ->_message( $colour, $label, $line , $reason )

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Paludis/UseCleaner.pm
use strict;
use warnings;

package Paludis::UseCleaner;
BEGIN {
  $Paludis::UseCleaner::AUTHORITY = 'cpan:KENTNL';
}
{
  $Paludis::UseCleaner::VERSION = '0.01000308';
}

# ABSTRACT: Remove cruft from your use.conf


use Moose;
use MooseX::Types::Moose qw( :all );
use MooseX::Types::Perl qw( :all );
use Cave::Wrapper;
use namespace::autoclean -also => qr/^__/;
use IO::Handle;
use Class::Load 0.06 qw( load_class );
use Moose::Util::TypeConstraints qw( class_type );
use MooseX::Has::Sugar;


has 'input' => ( isa => GlobRef, rw, required );


has 'output' => ( isa => GlobRef, rw, required );


has 'rejects' => ( isa => GlobRef, rw, required );


has 'debug' => ( isa => GlobRef, rw, required );


has 'dot_trace' => ( isa => GlobRef, rw, required );


has 'display_ui' => ( isa => Object, rw, lazy_build );


has 'display_ui_class' => ( isa => ModuleName, rw, lazy_build );


has 'display_ui_generator' => ( isa => CodeRef, rw, lazy_build );


sub do_work {

  my ($self) = shift;
  my $cave = Cave::Wrapper->new();

  $self->dot_trace->autoflush(1);

  while ( defined( my $line = $self->input->getline ) ) {

    my $lineno = $self->input->input_line_number;

    my (@tokens) = __tokenize($line);

    if ( __is_empty_line(@tokens) ) {
      $self->output->print($line);
      $self->display_ui->skip_empty( $lineno, $line );
      next;
    }
    if ( __is_star_rule(@tokens) ) {
      $self->output->print($line);
      $self->display_ui->skip_star( $lineno, $line );
      next;
    }
    $self->display_ui->dot_trace();

    my ( $spec, $use, $extras ) = __tokenparse(@tokens);

    $self->display_ui->full_rule( $spec, $use, $extras );

    my @packages = $cave->print_ids( '-m', $spec );

    if ( not @packages ) {
      $self->display_ui->nomatch( $lineno, $line );
      $self->rejects->print($line);
      next;
    }

    $self->output->print($line);
  }
  return;
}


sub __tokenize {
  my $line = shift;
  $line =~ s/#.*$//;
  return split /\s+/, $line;
}


## no critic (RequireArgUnpacking)

sub __is_empty_line {
  return not @_;
}


## no critic (RequireArgUnpacking)

sub __is_star_rule {
  return $_[0] =~ /\*/;
}


sub __tokenparse {
  my @tokens   = @_;
  my $spec     = shift @tokens;
  my @useflags = __extract_flags( \@tokens );
  my %extras;
  while ( defined( my $current = __extract_label( \@tokens ) ) ) {
    $extras{$current} = [ __extract_flags( \@tokens ) ];
  }
  return ( $spec, \@useflags, \%extras );
}


## no critic (ProhibitDoubleSigils)

sub __extract_flags {
  my $in = shift;
  my @out;
  while ( exists $in->[0] && $in->[0] !~ /^([A-Z_]+):$/ ) {
    push @out, shift @$in;
  }
  return @out;
}


## no critic (ProhibitDoubleSigils)
sub __extract_label {
  my $in = shift;
  return if not exists $in->[0];
  return if not $in->[0] =~ /^([A-Z_]+):$/;
  my $result = $1;
  shift @$in;
  return $result;
}


sub _build_display_ui_class {
  return 'Paludis::UseCleaner::ConsoleUI';
}


sub _build_display_ui_generator {
  my $self = shift;
  return sub {
    load_class( $self->display_ui_class );
    return $self->display_ui_class->new(
      fd_debug     => $self->debug,
      fd_dot_trace => $self->dot_trace,
    );
  };
}


sub _build_display_ui {
  my $self = shift;
  return $self->display_ui_generator()->($self);
}

no Moose;
no Moose::Util::TypeConstraints;

__PACKAGE__->meta->make_immutable;

1;

__END__

=pod

=head1 NAME

Paludis::UseCleaner - Remove cruft from your use.conf

=head1 VERSION

version 0.01000308

=head1 SYNOPSIS

This module handles the core behaviour of the Use Cleaner, to be consumed inside other applications.

For a "Just Use it" interface, you want L<paludis-usecleaner.pl> and L<Paludis::UseCleaner::App>

    my $cleaner = Paludis::UseCleaner->new(
        input     => somefd,
        output    => somefd,
        rejects   => somefd,
        debug     => fd_for_debugging
        dot_trace => fd_for_dot_traces,
      ( # Optional
        display_ui => $object_to_handle_debug_messages
        display_ui_class => $classname_to_construct_a_display_ui
        display_ui_generator => $coderef_to_generate_object_for_display_ui
      )
    );

    $cleaner->do_work();

=head1 METHODS

=head2 do_work

    $cleaner->do_work();

Executes the various transformations and produces the cleaned output from the input.

=head1 ATTRIBUTES

=head2 input

    $cleaner->input( \*STDIN );
    $cleaner->input( $read_fh );

=head2 output

    $cleaner->output( \*STDOUT );
    $cleaner->output( $write_fh );

=head2 rejects

    $cleaner->rejects( \*STDOUT );
    $cleaner->rejects( $write_fh );

=head2 debug

    $cleaner->debug( \*STDERR );
    $cleaner->debug( $write_fh );

=head2 dot_trace

    $cleaner->dot_trace( \*STDERR );
    $cleaner->dot_trace( $write_fh );

=head2 display_ui

    $cleaner->display_ui( $object );

=head2 display_ui_class

    $cleaner->display_ui_class( 'Some::Class::Name' );

=head2 display_ui_generator

    $cleaner->display_ui_generator( sub {
        my $self = shift;
        ....
        return $object;
    });

=head1 PRIVATE METHODS

=head2 __tokenize

    my @line = __tokenize( $line );

B<STRIPPED>: This method is made invisible to outside code after compile.

=head2 __is_empty_line

    if( __is_empty_line(@line) ){ }

B<STRIPPED>: This method is made invisible to outside code after compile.

=head2 __is_star_rule

    if( __is_star_rule(@line) ){ }

B<STRIPPED>: This method is made invisible to outside code after compile.

=head2 __tokenparse

    my ( $spec, $use, $extras ) = __tokenparse( @line );

B<STRIPPED>: This method is made invisible to outside code after compile.

=head2 __extract_flags

    my ( @flags ) = __extract_flags( \@tokens );

B<STRIPPED>: This method is made invisible to outside code after compile.

=head2 __extract_label

    my ( $label ) = __extract_label( \@tokens );

B<STRIPPED>: This method is made invisible to outside code after compile.

=head2 _build_display_ui_class

    my $class = $cleaner->_build_display_ui_class();

=head2 _build_display_ui_generator

    my $generator  $cleaner->_build_display_ui_generator();

=head2 _build_display_ui

    my $object = $cleaner->_build_display_ui();

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
